FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageQueueClient.cs

```diff
@@ -0,0 +1,682 @@
+        {
+            _logger.LogInformation($"Start to cancel job {jobId}.");
+
+            // step 1:Load reverse index entity.
+            var reverseIndexEntity = await GetJobReverseIndexEntityByIdAsync(queueType, jobId, cancellationToken);
+
+            // step 2: get jobInfo entity
+            var jobInfoEntity = (await _azureJobInfoTableClient.GetEntityAsync<JobInfoEntity>(reverseIndexEntity.JobInfoEntityPartitionKey, reverseIndexEntity.JobInfoEntityRowKey, cancellationToken: cancellationToken)).Value;
+
+            // step 3: cancel job
+            CancelJobInternal(jobInfoEntity);
+
+            // step 4: update job info entity to table.
+            await _azureJobInfoTableClient.UpdateEntityAsync(jobInfoEntity, jobInfoEntity.ETag, cancellationToken: cancellationToken);
+
+            _logger.LogInformation($"Cancel job {jobId} successfully.");
+        }
+
+        public async Task CompleteJobAsync(JobInfo jobInfo, bool requestCancellationOnFailure, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation($"Start to complete job {jobInfo.Id}.");
+
+            Trace.Assert(
+                jobInfo.Result.Length * sizeof(char) < MaximumSizeOfAnEntityInBytes,
+                "The maximum size of a single table entity is 1MB, the size of result is larger than 1MB.");
+
+            // step 1: check version
+            var retrievedJobInfo = await GetJobByIdAsync(jobInfo.QueueType, jobInfo.Id, false, cancellationToken);
+
+            if (retrievedJobInfo.Version != jobInfo.Version)
+            {
+                _logger.LogError($"Job {jobInfo.Id} precondition failed, version does not match.");
+                throw new JobNotExistException($"Job {jobInfo.Id} precondition failed, version does not match.");
+            }
+
+            // step 2: get shouldCancel
+            var shouldCancel = retrievedJobInfo.CancelRequested;
+
+            // step 3: update status
+            // Reference: https://github.com/microsoft/fhir-server/blob/e1117009b6db995672cc4d31457cb3e6f32e19a3/src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/PutJobStatus.sql#L16
+            var jobInfoEntity = ((TJobInfo)jobInfo).ToTableEntity();
+            if (jobInfoEntity.Status == (int)JobStatus.Failed)
+            {
+                jobInfoEntity.Status = (int)JobStatus.Failed;
+            }
+            else if (shouldCancel)
+            {
+                jobInfoEntity.Status = (int)JobStatus.Cancelled;
+            }
+            else
+            {
+                jobInfoEntity.Status = (int) JobStatus.Completed;
+            }
+
+            // step 4: update job info entity to table
+            await _azureJobInfoTableClient.UpdateEntityAsync(jobInfoEntity, ETag.All, cancellationToken: cancellationToken);
+
+            // step 5: get job lock entity to get job message id and job pop receipt
+            var jobLockEntity = (await _azureJobInfoTableClient.GetEntityAsync<JobLockEntity>(
```

Comment:
Merge get operation?

Reply:
Do you mean merging with the `GetJobByIdAsync` function?

Reply:
To merge get operation, we need to use generic TableEntity instead of JobInfoEntity and JobLockEntity, a [new pr](https://github.com/microsoft/FHIR-Analytics-Pipelines/pull/153) is created for it.